### DISCUSSION TITLE
Hyperbolic Embedding - replacing dot product with manifold distance

### DISCUSSION THREAD

**Discussion by tsteffek** [geoopt/geoopt#182] [Q&A]
Hi!

I'm currently trying to reimplement/modify Chatterjee's [Joint Learning of Hyperbolic Label Embeddings for Hierarchical Multi-label Classification](https://aclanthology.org/2021.eacl-main.247/), [github link](https://github.com/soumyac1999/hyperbolic-label-emb-for-hmc).

Their approach combines two losses:
 - they min/max the distance between related/unrelated labels in their embedding using their own [Poincaré distance](https://github.com/soumyac1999/hyperbolic-label-emb-for-hmc/blob/3e3ab4c2d17521fbc7414d477dcca630b3afc5ea/poincare_utils.py#L22) (the only riemannian part of the whole project)
 - they compute the dot product of their label and document embedding (created using a TextCNN), followed by Sigmoid + BCE

Ignoring the first loss for now, I'm trying to improve on their approach by replacing the dot product with the Poincaré distance (embedding size of 20, so hopefully curse of dimensionality doesn't hit too bad).

Basically this:
![image](https://user-images.githubusercontent.com/28529410/129900747-dfcd489d-ea13-479b-98df-47e05ab615d7.png)
My modified setup uses a BERT model, a handwritten `ManifoldEmbedding` module, which uses ManifoldParameters instead of normal ones, and either a `MobiusLinear` (following your example) + projection or a normal `Linear` layer.

However, the loss is very erratic and precision and recall stay at 0 even when trying to overfit just a single datapoint (same result with varying learning rates, RiemannianAdam, 64-bit):
![image](https://user-images.githubusercontent.com/28529410/129902010-26f79472-445e-4806-baad-393adc63dfea.png)
//EDIT: just checked this highly erratic one is with a normal `Linear` layer, the Projection + Mobius is pretty straight, but still remains at 0 precision/recall

What do you think of this modification? For me and my supervisor the approach seems to be sound, but we can't figure out why it won't learn. I'm especially unsure about when to use projections.

**Comment by ferrine** [comment]
Do you have a gist to be more specific on the implementation?

**Reply by tsteffek** [reply]
Here you go, thanks for taking the time: https://gist.github.com/tsteffek/26fdca9caab9aa75e4887f159326e375

I've tried to remove all irrelevant parts like logging and some inheritance, so it might not work as it is right now, but the important parts should be there.
Also, in case you're unfamiliar with PyTorch Lightning: it's basically a nice PyTorch wrapper which manages a lot of the housekeeping and enforces the specific functions like `configure_optimizers`, `training_step`, and so on. I think it's pretty straightforward, hit me up if anything's unclear. (Also I've merged test and val step in my BaseLightningModel)

**Reply by ferrine** [reply]
I work with lightning too much))

**Comment by ferrine** [comment]
Great, I'll have a look this weekend

**Comment by ferrine** [comment]
Have you tried dist2 (squared distance)?  It has a smoother loss landscape. Also, small learning rates worked better for me long time ago. Manifold optimization is too tricky

**Comment by ferrine** [comment]
Otherwise, the implementation does not seem to have immediate issues I can spot.

**Comment by ferrine** [comment]
After a second pass (I was too curious and comment a lot here, but still). I think you use the projection non efficiently. BERT should have a lot of dimensions and they have to be compressed before the actual projection. Moreover BERT does not have the fancy tree-like structure embeddings. I'd replace BETR-EXP0-MLinear with BERT-MLP-EXP0, where MLP is 1-2 Dense Layers.

**Reply by tsteffek** [reply]
The Dense Layers here being non-mobius, right?
Thanks again for your time, I'll try out your suggestions and keep you updated.

**Reply by tsteffek** [reply]
So Linear-Relu-Linear-exp0 as MLP didn't work, same result as before. Only tried overfitting with 1 batch / batch_size 4, but since it didn't manage to get away from 0 precision/recall after 500 steps I guess that means there's something fundamentally wrong.

It's the same result with `torch.cdist` as euclidean approach. Only dot works. So maybe it's a problem with any distance metrics. But Chatterjee's approach of using dot for classification while simultaneously trying to enforce a hyperbolic embedding structure seems counterintuitive to me; after all, the dot knows nothing about the laws of the Poincare Model and basically uses its own "distance metric". Or am I missing something here? What's your take on this @ferrine?
